core/trino-main/src/main/java/io/trino/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -370,7 +371,9 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
         // NOTE: buffers are allowed to be created in the FINISHED state because destroy() can move to the finished state
         // without a clean "no-more-buffers" message from the scheduler.  This happens with limit queries and is ok because
         // the buffer will be immediately destroyed.
-        checkState(stateMachine.getState().canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
+        BufferState state = stateMachine.getState();
+        // buffer may become aborted while the final output buffers are being set
+        checkState(state == ABORTED || state.canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
```

Comment:
shouldn't you destroy buffer immediately if state is `aborted` like below:
```
        if (stateMachine.getState() == FINISHED) {
            buffer.destroy();
        }
```
?

Reply:
This is something that is very non obvious in OutputBuffers implementation. We destroy buffers only if an upstream task finished successfully as "destroy" usually propagates a "success" signal to a downstream task. However if an upstream task is failed we don't destroy buffers and wait for coordinator to kill downstream tasks.